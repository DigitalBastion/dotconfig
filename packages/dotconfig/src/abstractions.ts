import type { ConfigurationTypeSymbol } from "./constants.js";

export interface IConfigurationBuilder {
  /**
   * Gets a key/value collection that can be used to share data between the {@link IConfigurationBuilder}
   * and the registered {@link IConfigurationSource}'s.
   */
  readonly properties: Map<string, unknown>;

  readonly sources: IConfigurationSource[];

  add(source: IConfigurationSource): this;
  build(): Promise<IConfigurationRoot>;
}

export interface IConfiguration extends Iterable<[string, string | null]> {
  [ConfigurationTypeSymbol]: "root" | "section";
  get(key: string): string | null;
  set(key: string, value: string | null): void;
  /**
   * Gets a configuration sub-section with the specified key.
   * @param key The key of the configuration section.
   *
   * @remarks
   *
   * This method will never return `null`. If no matching sub-section is found with the specified key, an empty {@link IConfigurationSection} will be returned.
   */
  getSection(key: string): IConfigurationSection;
  /**
   * Gets a configuration subsection with the specified key.
   * @param key The key of the configuration section.
   *
   * @remarks
   *
   * If no matching sub-section is found with the specified key, an exception is raised.
   */
  getRequiredSection(key: string): IConfigurationSection;
  /**
   * Gets the immediate descendant configuration sub-sections.
   */
  getChildren(key?: string): IConfigurationSection[];
  getReloadToken(): IChangeToken;
}

export interface IConfigurationRoot extends IConfiguration {
  [ConfigurationTypeSymbol]: "root";
  /**
   * Gets the {@link IConfigurationProvider} providers for this configuration.
   */
  readonly providers: IConfigurationProvider[];

  /**
   * Forces the configuration values to be reloaded from the underlying {@link IConfigurationProvider} providers.
   */
  reload(): Promise<void>;
}

export interface IConfigurationSection extends IConfiguration {
  [ConfigurationTypeSymbol]: "section";
  /**
   * Gets the key this section occupies in its parent.
   */
  readonly key: string;
  /**
   * Gets the full path to this section within the {@link IConfiguration}.
   */
  readonly path: string;

  value: string | null;

  exists(): boolean;
}

export interface IConfigurationProvider {
  get(key: string): string | null | undefined;
  set(key: string, value: string | null): void;
  /**
   * Returns the immediate descendant configuration keys for a given parent path based on the data of this
   * {@link IConfigurationProvider} and the set of keys returned by all the preceding
   * {@link IConfigurationProvider} providers.
   *
   * @param earlierKeys The child keys returned by the preceding providers for the same parent path.
   * @param parentPath The parent path.
   */
  getChildKeys(earlierKeys: string[], parentPath?: string): string[];
  /**
   * Loads configuration values from the source represented by this {@link IConfigurationProvider}.
   */
  load(): Promise<void>;

  getReloadToken(): IChangeToken | null;
}

export interface IConfigurationSource {
  build(builder: IConfigurationBuilder): Promise<IConfigurationProvider>;
}

export interface IChangeToken {
  /**
   * Gets a value that indicates if a change has occurred.
   */
  hasChanged: boolean;
  /**
   * Indicates if this token will proactively raise callbacks. If `false`, the token consumer must poll {@link hasChanged} to detect changes.
   */
  activeChangeCallbacks: boolean;
  /**
   * Registers for a callback that will be invoked when the entry has changed.
   * {@link hasChanged} MUST be set before the callback is invoked.
   *
   * @param callback  The callback to invoke.
   * @returns A function that can be used to unregister the callback.
   */
  registerChangeCallback: (callback: (state?: unknown) => void, state?: unknown) => Disposable;
}
