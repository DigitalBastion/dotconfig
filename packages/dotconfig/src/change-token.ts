import type { IChangeToken } from "./abstractions.js";

export function createChangeToken(changeTokenProducer: () => IChangeToken | null, changeTokenConsumer: () => void) {
  return new ChangeTokenRegistration(changeTokenProducer, (callback) => callback(), changeTokenConsumer);
}

export function createChangeTokenWithState<TState>(
  changeTokenProducer: () => IChangeToken | null,
  changeTokenConsumer: (state: TState) => void,
  state: TState,
) {
  return new ChangeTokenRegistration(changeTokenProducer, changeTokenConsumer, state);
}

class ChangeTokenRegistration<TState> implements Disposable {
  public constructor(
    changeTokenProducer: () => IChangeToken | null,
    changeTokenConsumer: (state: TState) => void,
    state: TState,
  ) {
    this.#changeTokenProducer = changeTokenProducer;
    this.#changeTokenConsumer = changeTokenConsumer;
    this.#state = state;

    const token = changeTokenProducer();

    this.registerChangeTokenCallback(token);
  }

  #changeTokenProducer: () => IChangeToken | null;
  #changeTokenConsumer: (state: TState) => void;
  #state: TState;
  #disposable: Disposable | null = null;
  #disposedSentinel: Disposable = new NoopDisposable();

  private onChangeTokenFired() {
    const token = this.#changeTokenProducer();

    try {
      this.#changeTokenConsumer(this.#state);
    } finally {
      this.registerChangeTokenCallback(token);
    }
  }

  private registerChangeTokenCallback(token: IChangeToken | null) {
    if (token == null) {
      return;
    }

    const registration = token.registerChangeCallback(
      (s) => (s as ChangeTokenRegistration<TState>).onChangeTokenFired(),
      this,
    );

    if (token.hasChanged && token.activeChangeCallbacks) {
      registration[Symbol.dispose]();
      return;
    }

    this.setDisposable(registration);
  }

  private setDisposable(disposable: Disposable) {
    const current = this.#disposable;

    // If Dispose was called, then immediately dispose the disposable
    if (current === this.#disposedSentinel) {
      disposable[Symbol.dispose]();
      return;
    }

    // Otherwise, try to update the disposable.
    const previous = current;
    this.#disposable = disposable;

    if (previous === this.#disposedSentinel) {
      // The subscription was disposed so we dispose immediately and return.
      disposable[Symbol.dispose]();
    } else if (previous === current) {
      // We successfully assigned the _disposable field to disposable.
      return;
    } else {
      // Sets can never overlap with other SetDisposable calls so we should never get into this situation.
      throw new Error("Somebody else set the _disposable field.");
    }
  }

  [Symbol.dispose](): void {
    this.#disposable?.[Symbol.dispose]();
    this.#disposable = this.#disposedSentinel;
  }
}

class NoopDisposable implements Disposable {
  [Symbol.dispose](): void {
    // Noop
  }
}
