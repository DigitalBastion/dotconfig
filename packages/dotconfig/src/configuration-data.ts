export class ConfigurationData extends Map<string, string | null> {
    #caseInsensitiveMap = new Map<string, string>();

    public get(key: string) {
        const casedKey = this.#caseInsensitiveMap.get(key.toLowerCase());
        return casedKey ? super.get(casedKey) : undefined;
    }

    public set(key: string, value: string | null) {
        this.#caseInsensitiveMap.set(key.toLowerCase(), key);

        return super.set(key, value);
    }

    public delete(key: string) {
        const casedKey = this.#caseInsensitiveMap.get(key.toLowerCase());
        if (casedKey == null) {
            return false;
        }

        this.#caseInsensitiveMap.delete(key.toLowerCase());
        return super.delete(casedKey);
    }

    public has(key: string) {
        return this.#caseInsensitiveMap.has(key.toLowerCase());
    }
}
