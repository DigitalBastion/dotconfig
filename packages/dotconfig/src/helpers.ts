import type { IConfiguration } from "./abstractions.js";
import { isConfigurationSection } from "./utils.js";

/**
 * Get the iterator of key-value pairs for the configuration.
 * @param configuration The configuration to iterate.
 * @param makePathsRelative `true` to trim the current configuration's path from the front of the returned child keys.
 */
export function* iterateConfigurationEntries(
  configuration: IConfiguration,
  makePathsRelative = false,
): Iterable<readonly [string, string | null]> {
  const stack: IConfiguration[] = [configuration];
  const prefixLength = makePathsRelative && isConfigurationSection(configuration) ? configuration.path.length + 1 : 0;

  while (stack.length > 0) {
    // biome-ignore lint/style/noNonNullAssertion: Already checked in the while loop.
    const config = stack.pop()!;

    // Don't include the sections value if we are removing paths, since it will be an empty key
    if (isConfigurationSection(config) && (!makePathsRelative || config !== configuration)) {
      yield [config.path.substring(prefixLength), config.value];
    }

    for (const child of config.getChildren()) {
      stack.push(child);
    }
  }
}


export function dispose(instance: unknown): void {
  if (typeof instance !== "object" || instance === null) {
    return;
  }

  (instance as Disposable)[Symbol.dispose]?.();
}
