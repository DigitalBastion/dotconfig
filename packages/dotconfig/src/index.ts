export * from "./configuration-builder.js";
export * from "./helpers.js";
