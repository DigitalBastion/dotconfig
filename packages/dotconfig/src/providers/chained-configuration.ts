import type { IConfiguration, IConfigurationProvider, IConfigurationSource } from "../abstractions.js";
import { compare } from "../configuration-path.js";

export class ChainedConfigurationSource implements IConfigurationSource {
  constructor(configuration: IConfiguration) {
    this.#configuration = configuration;
  }

  #configuration: IConfiguration;

  public get configuration() {
    return this.#configuration;
  }

  public async build() {
    return new ChainedConfigurationProvider(this);
  }
}

export class ChainedConfigurationProvider implements IConfigurationProvider {
  constructor(source: ChainedConfigurationSource) {
    this.#source = source;
  }

  #source: ChainedConfigurationSource;

  public get(key: string): string | null | undefined {
    return this.#source.configuration.get(key);
  }

  public set(key: string, value: string | null): void {
    this.#source.configuration.set(key, value);
  }

  public getChildKeys(earlierKeys: string[], parentPath?: string): string[] {
    const section = parentPath == null ? this.#source.configuration : this.#source.configuration.getSection(parentPath);
    const keys: string[] = [...earlierKeys];

    for (const child of section.getChildren()) {
      keys.push(child.key);
    }

    keys.sort(compare);
    return keys;
  }

  public load(): Promise<void> {
    return Promise.resolve();
  }

  public getReloadToken() {
    return this.#source.configuration.getReloadToken();
  }
}
