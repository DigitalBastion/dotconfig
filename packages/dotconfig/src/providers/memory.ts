import type { IConfigurationProvider } from "../abstractions.js";
import { ConfigurationProvider } from "../configuration-provider.js";
import { ConfigurationSource } from "../configuration-source.js";

export class MemoryConfigurationSource extends ConfigurationSource {
  constructor(initialData: Map<string, string | null>) {
    super();
    this.#initialData = initialData;
  }

  #initialData: Map<string, string | null>;

  public get initialData(): Map<string, string | null> {
    return this.#initialData;
  }

  public async build(): Promise<IConfigurationProvider> {
    const provider = new MemoryConfigurationProvider(this);
    await provider.load();

    return provider;
  }
}

export class MemoryConfigurationProvider extends ConfigurationProvider {
  constructor(source: MemoryConfigurationSource) {
    super();
    this.#source = source;
  }

  #source: MemoryConfigurationSource;

  public override load(): Promise<void> {
    for (const [key, value] of this.#source.initialData) {
      this.set(key, value);
    }

    return Promise.resolve();
  }
}
