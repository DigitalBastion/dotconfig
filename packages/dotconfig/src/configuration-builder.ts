import type { IConfigurationBuilder, IConfigurationProvider, IConfigurationSource } from "./abstractions.js";
import { ConfigurationRoot } from "./configuration-root.js";
import { EnvironmentSource } from "./providers/environment.js";
import { MemoryConfigurationSource } from "./providers/memory.js";

export class ConfigurationBuilder implements IConfigurationBuilder {
    #sources: IConfigurationSource[] = [];
    #properties = new Map<string, unknown>();

    public get sources() {
        return this.#sources;
    }

    public get properties() {
        return this.#properties;
    }

    public add(source: IConfigurationSource): this {
        this.#sources.push(source);
        return this;
    }

    public addEnvironmentVariables(delimiter?: string): this {
        this.add(new EnvironmentSource({ delimiter }));
        return this;
    }

    public addMemoryCollection(collection: Map<string, string | null>): this {
        this.add(new MemoryConfigurationSource(collection));
        return this;
    }

    public async build() {
        const providerPromises: Array<Promise<IConfigurationProvider>> = this.#sources.map(source => source.build(this));
        const providers = await Promise.all(providerPromises);
        return new ConfigurationRoot(providers);
    }
}