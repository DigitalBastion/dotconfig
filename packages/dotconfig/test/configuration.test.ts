import { test, expect, describe } from "bun:test";
import { MemoryConfigurationSource } from "../src/providers/memory";
import { ConfigurationBuilder } from "../src/configuration-builder";
import { ChainedConfigurationProvider, ChainedConfigurationSource } from "../src/providers/chained-configuration";
import { iterateConfigurationEntries } from "../src/helpers";

test("Load and combine key value pairs from different configuration providers", async () => {
    const dic1 = new Map<string, string>([
        ["Mem1:KeyInMem1", "ValueInMem1"]
    ]);
    const dic2 = new Map<string, string>([
        ["Mem2:KeyInMem2", "ValueInMem2"]
    ]);
    const dic3 = new Map<string, string>([
        ["Mem3:KeyInMem3", "ValueInMem3"]
    ]);

    const memConfigSrc1 = new MemoryConfigurationSource(dic1);
    const memConfigSrc2 = new MemoryConfigurationSource(dic2);
    const memConfigSrc3 = new MemoryConfigurationSource(dic3);

    const configurationBuilder = new ConfigurationBuilder();

    // Act
    configurationBuilder.add(memConfigSrc1);
    configurationBuilder.add(memConfigSrc2);
    configurationBuilder.add(memConfigSrc3);

    const config = await configurationBuilder.build();

    const memVal1 = config.get("mem1:keyinmem1");
    const memVal2 = config.get("Mem2:KeyInMem2");
    const memVal3 = config.get("MEM3:KEYINMEM3");

    // Expect
    expect(configurationBuilder.sources).toContain(memConfigSrc1);
    expect(configurationBuilder.sources).toContain(memConfigSrc2);
    expect(configurationBuilder.sources).toContain(memConfigSrc3);

    expect(memVal1).toBe("ValueInMem1");
    expect(memVal2).toBe("ValueInMem2");
    expect(memVal3).toBe("ValueInMem3");

    expect(config.get("mem1:keyinmem1")).toBe("ValueInMem1");
    expect(config.get("Mem2:KeyInMem2")).toBe("ValueInMem2");
    expect(config.get("MEM3:KEYINMEM3")).toBe("ValueInMem3");
    expect(config.get("NotExist")).toBeNull();
});

describe("GetChildKeys", () => {
    test("Can chain empty keys", async () => {
        const input = new Map<string, string>();
        for (let i = 0; i < 1000; i++) {
            input.set(" ".repeat(i), "");
        }

        const configurationRoot = await new ConfigurationBuilder()
            .add(new MemoryConfigurationSource(input))
            .build();

        const chainedConfigurationSource = new ChainedConfigurationSource(configurationRoot);

        const chainedConfiguration = new ChainedConfigurationProvider(chainedConfigurationSource);
        const childKeys = chainedConfiguration.getChildKeys([]);

        expect(childKeys.length).toBe(1000);
        expect(childKeys[0]).toBe("");
    });

    test.skip("Can chain key with no delimiter", () => { });
});

test("Can chain configuration", async () => {
    // Arrange
    const dic1 = new Map<string, string>([
        ["Mem1:KeyInMem1", "ValueInMem1"]
    ]);
    const dic2 = new Map<string, string>([
        ["Mem2:KeyInMem2", "ValueInMem2"]
    ]);
    const dic3 = new Map<string, string>([
        ["Mem3:KeyInMem3", "ValueInMem3"]
    ]);

    const memConfigSrc1 = new MemoryConfigurationSource(dic1);
    const memConfigSrc2 = new MemoryConfigurationSource(dic2);
    const memConfigSrc3 = new MemoryConfigurationSource(dic3);

    const configurationBuilder = new ConfigurationBuilder();

    // Act
    configurationBuilder.add(memConfigSrc1);
    configurationBuilder.add(memConfigSrc2);
    configurationBuilder.add(memConfigSrc3);

    const config = await configurationBuilder.build();

    const chained = await new ConfigurationBuilder()
        .add(new ChainedConfigurationSource(config))
        .build();

    const memVal1 = chained.get("mem1:keyinmem1");
    const memVal2 = chained.get("Mem2:KeyInMem2");
    const memVal3 = chained.get("MEM3:KEYINMEM3");

    // Assert
    expect(memVal1).toBe("ValueInMem1");
    expect(memVal2).toBe("ValueInMem2");
    expect(memVal3).toBe("ValueInMem3");

    expect(chained.get("NotExist")).toBeNull();
});

test.each([true, false])("Chained getConfigurationEntries flattens into dictionary (removePath: %s)", async (removePath) => {
    // Arrange
    const dic1 = new Map([
        ["Mem1", "Value1"],
        ["Mem1:", "NoKeyValue1"],
        ["Mem1:KeyInMem1", "ValueInMem1"],
        ["Mem1:KeyInMem1:Deep1", "ValueDeep1"]
    ]);
    const dic2 = new Map([
        ["Mem2", "Value2"],
        ["Mem2:", "NoKeyValue2"],
        ["Mem2:KeyInMem2", "ValueInMem2"],
        ["Mem2:KeyInMem2:Deep2", "ValueDeep2"]
    ]);
    const dic3 = new Map([
        ["Mem3", "Value3"],
        ["Mem3:", "NoKeyValue3"],
        ["Mem3:KeyInMem3", "ValueInMem3"],
        ["Mem3:KeyInMem3:Deep3", "ValueDeep3"]
    ]);

    const memConfigSrc1 = new MemoryConfigurationSource(dic1);
    const memConfigSrc2 = new MemoryConfigurationSource(dic2);
    const memConfigSrc3 = new MemoryConfigurationSource(dic3);

    const configurationBuilder = new ConfigurationBuilder();

    // Act
    configurationBuilder.add(memConfigSrc1);
    configurationBuilder.add(memConfigSrc2);
    const config = await new ConfigurationBuilder()
        .add(new ChainedConfigurationSource(await configurationBuilder.build()))
        .add(memConfigSrc3)
        .build();

    const dict = Object.fromEntries(iterateConfigurationEntries(config, removePath));

    // Assert
    expect(dict).toMatchSnapshot();
});

test("New configuration provider overrides old one when key is duplicated", async () => {
    // Arrange
    const dic1 = new Map([
        ["Key1:Key2", "ValueInMem1"]
    ]);
    const dic2 = new Map([
        ["Key1:Key2", "ValueInMem2"]
    ]);

    const memConfigSrc1 = new MemoryConfigurationSource(dic1);
    const memConfigSrc2 = new MemoryConfigurationSource(dic2);

    const configurationBuilder = new ConfigurationBuilder();

    // Act
    configurationBuilder.add(memConfigSrc1);
    configurationBuilder.add(memConfigSrc2);

    const config = await configurationBuilder.build();

    // Assert
    expect(config.get("Key1:Key2")).toBe("ValueInMem2");
});
