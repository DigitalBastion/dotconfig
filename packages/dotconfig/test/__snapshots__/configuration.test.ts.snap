// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`Chained getConfigurationEntries flattens into dictionary (removePath: %s) 1`] = `
{
  "Mem1": "Value1",
  "Mem1:": "NoKeyValue1",
  "Mem1:KeyInMem1": "ValueInMem1",
  "Mem1:KeyInMem1:Deep1": "ValueDeep1",
  "Mem2": "Value2",
  "Mem2:": "NoKeyValue2",
  "Mem2:KeyInMem2": "ValueInMem2",
  "Mem2:KeyInMem2:Deep2": "ValueDeep2",
  "Mem3": "Value3",
  "Mem3:": "NoKeyValue3",
  "Mem3:KeyInMem3": "ValueInMem3",
  "Mem3:KeyInMem3:Deep3": "ValueDeep3",
}
`;

exports[`Chained getConfigurationEntries flattens into dictionary (removePath: %s) 2`] = `
{
  "Mem1": "Value1",
  "Mem1:": "NoKeyValue1",
  "Mem1:KeyInMem1": "ValueInMem1",
  "Mem1:KeyInMem1:Deep1": "ValueDeep1",
  "Mem2": "Value2",
  "Mem2:": "NoKeyValue2",
  "Mem2:KeyInMem2": "ValueInMem2",
  "Mem2:KeyInMem2:Deep2": "ValueDeep2",
  "Mem3": "Value3",
  "Mem3:": "NoKeyValue3",
  "Mem3:KeyInMem3": "ValueInMem3",
  "Mem3:KeyInMem3:Deep3": "ValueDeep3",
}
`;

exports[`Get configuration nested object 1`] = `
{
  "0": "NoKeyValue1",
  "Deep1": "ValueDeep1",
  "KeyInMem1": "ValueInMem1",
  "Mem1": "Value1",
}
`;
